src/value.rs

```diff
@@ -79,50 +83,126 @@ impl<'gc> Value<'gc> {
-        }
+        bin_op(
+            self,
+            other,
+            |a, b| Value::Integer(a.wrapping_add(b)),
+            |a, b| Value::Number(a + b),
+        )
     }
 
     pub fn subtract(self, other: Value<'gc>) -> Option<Value<'gc>> {
-        match (self, other) {
-            (Value::Integer(a), Value::Integer(b)) => Some(Value::Integer(a.wrapping_sub(b))),
-            (Value::Number(a), Value::Number(b)) => Some(Value::Number(a - b)),
-            (Value::Integer(a), Value::Number(b)) => Some(Value::Number(a as f64 - b)),
-            (Value::Number(a), Value::Integer(b)) => Some(Value::Number(a - b as f64)),
-            _ => None,
-        }
+        bin_op(
+            self,
+            other,
+            |a, b| Value::Integer(a.wrapping_sub(b)),
+            |a, b| Value::Number(a - b),
+        )
     }
 
     pub fn multiply(self, other: Value<'gc>) -> Option<Value<'gc>> {
-        match (self, other) {
-            (Value::Integer(a), Value::Integer(b)) => Some(Value::Integer(a.wrapping_mul(b))),
-            (Value::Number(a), Value::Number(b)) => Some(Value::Number(a * b)),
-            (Value::Integer(a), Value::Number(b)) => Some(Value::Number(a as f64 * b)),
-            (Value::Number(a), Value::Integer(b)) => Some(Value::Number(a * b as f64)),
-            _ => None,
+        bin_op(
+            self,
+            other,
+            |a, b| Value::Integer(a.wrapping_mul(b)),
+            |a, b| Value::Number(a * b),
+        )
+    }
+
+    /// This operation always returns a Number, even when called with Integer arguments.
+    pub fn float_divide(self, other: Value<'gc>) -> Option<Value<'gc>> {
+        bin_op(
+            self,
+            other,
+            |a, b| safe_div(a, b, |a, b| Value::Number(a as f64 / b as f64)),
+            |a, b| safe_div(a, b, |a, b| Value::Number(a / b)),
+        )
+    }
+
+    /// This operation returns an Integer only if both arguments are Integers.  Rounding is towards
+    /// negative infinity.
+    pub fn floor_divide(self, other: Value<'gc>) -> Option<Value<'gc>> {
+        bin_op(
+            self,
+            other,
+            |a, b| safe_div(a, b, |a, b| Value::Integer(a.wrapping_div(b))),
+            |a, b| safe_div(a, b, |a, b| Value::Number((a / b).floor())),
+        )
+    }
+
+    /// Computes the Lua modulus (`%`) operator.  This is unlike Rust's `%` operator which computes
+    /// the remainder.
+    pub fn modulo(self, other: Value<'gc>) -> Option<Value<'gc>> {
+        if self.to_integer().is_some() && other.to_integer() == Some(0) {
+            return None;
+        }
+
+        if self.to_number().is_some() && other.to_number() == Some(0.0) {
+            return Some(Value::Number(-f64::NAN));
         }
+
+        bin_op(
+            self,
+            other,
+            |a, b| safe_div(a, b, |a, b| Value::Integer(((a % b) + b) % b)),
+            |a, b| safe_div(a, b, |a, b| Value::Number(((a % b) + b) % b)),
+        )
```

Comment:
As per discussion, `safe_div` doesn't really buy much here, and this will potentially parse a number twice.